### PR TITLE
[20.09] Various bugfixes for reloading the toolbox

### PR DESCRIPTION
**Pull request by mvdbeek**
Fixes https://github.com/galaxyproject/galaxy/issues/10816 and some more things that were problematic.

### CODE REVIEW

**Comment by mvdbeek** [comment]
Test errors are unrelated.

File: lib/galaxy/tools/cache.py [hunk 1]
```diff
@@ -161,11 +161,11 @@ def cleanup(self):
         removed_tool_ids = []
         try:
             with self._lock:
-                persist_tool_document_cache = False
+                persist_callback = None
                 paths_to_cleanup = {(path, tool) for path, tool in self._tools_by_path.items() if self._should_cleanup(path)}
                 for config_filename, tool in paths_to_cleanup:
                     tool.remove_from_cache()
-                    persist_tool_document_cache = True
+                    persist_callback = tool.app.toolbox.persist_cache
```
Comment (nsoranzo) [inline lib/galaxy/tools/cache.py]: I preferred the previous version using a boolean to track if we need to persist, which is also slightly more efficient being in a loop?

File: lib/galaxy/tools/cache.py [hunk 2]
```diff
@@ -53,7 +53,7 @@ def _get_cache(self, flag='r', create_if_necessary=False):
                 if flag == 'r':
                     self._cache.flag = flag
             else:
-                self._cache = SqliteDict(self.cache_file, flag=flag, encode=encoder, decode=decoder, autocommit=False)
+                self._cache = SqliteDict(self.writeable_cache_file and self.writeable_cache_file.name or self.cache_file, flag=flag, encode=encoder, decode=decoder, autocommit=False)
```
Comment (nsoranzo) [inline lib/galaxy/tools/cache.py]: Can you split this in 2 statements for clarity?
Also the `... and ... or ` is a bit "Perlish", `self.writeable_cache_file.name if self.writeable_cache_file else self.cache_file` is more understandable?

File: lib/galaxy/tools/search/__init__.py [hunk 3]
```diff
@@ -90,11 +90,11 @@ def build_index(self, tool_cache, index_help=True):
...
         for indexed_tool_id in indexed_tool_ids:
             indexed_tool = tool_cache.get_tool_by_id(indexed_tool_id)
-            if not indexed_tool:
-                tool_ids_to_remove.add(indexed_tool_id)
-                continue
-            if not indexed_tool.is_latest_version and not indexed_tool.latest_version.hidden:
-                tool_ids_to_remove.add(indexed_tool_id)
+            if indexed_tool:
+                latest_version = indexed_tool.latest_version
+                if latest_version and latest_version.hidden:
```
Comment (nsoranzo) [inline lib/galaxy/tools/search/__init__.py:97]: Not sure if it's on purpose, but this seems functionally different from before, where we didn't remove an old version if the latest is hidden, xref. https://github.com/galaxyproject/galaxy/pull/10175#issuecomment-685001855

Comment (mvdbeek) [reply lib/galaxy/tools/search/__init__.py:97]: Are you sure ? My brain isn't working super-well at this hour, but I think this is equivalent, i've just inverted the continue.

File: lib/galaxy/tools/search/__init__.py [hunk 4]
```diff
@@ -90,11 +90,11 @@ def build_index(self, tool_cache, index_help=True):
         tool_ids_to_remove = (indexed_tool_ids - set(tool_cache._tool_paths_by_id.keys())).union(tool_cache._removed_tool_ids)
         for indexed_tool_id in indexed_tool_ids:
             indexed_tool = tool_cache.get_tool_by_id(indexed_tool_id)
-            if not indexed_tool:
-                tool_ids_to_remove.add(indexed_tool_id)
-                continue
-            if not indexed_tool.is_latest_version and not indexed_tool.latest_version.hidden:
-                tool_ids_to_remove.add(indexed_tool_id)
+            if indexed_tool:
+                latest_version = indexed_tool.latest_version
```
Comment (nsoranzo) [inline lib/galaxy/tools/search/__init__.py]:
```suggestion
            if indexed_tool:
                if indexed_tool.is_latest_version:
                    continue
                latest_version = indexed_tool.latest_version
```

File: lib/galaxy/tools/search/__init__.py (hunk 3 above)
Comment (nsoranzo) [reply lib/galaxy/tools/search/__init__.py:97]: The inverted condition (without the fix) would be:
```python
if indexed_tool.is_latest_version or indexed_tool.latest_version.hidden:
    continue
```

See my suggestion above.

**Review by jmchilton — APPROVED** [review]

**Review by nsoranzo — APPROVED** [review]

**Pull request merged**